packages/realm/src/Logger.ts

```diff
@@ -20,18 +20,137 @@ import { assert, binding } from "./internal";
+   * attention to efficiency.
+   */
   Debug = 2,
+  /**
+   * Same as 'Info', but prioritize completeness over minimalism.
+   */
   Detail = 3,
+  /**
+   * Reveal information about what is going on, but in a
+   * minimalistic fashion to avoid general overhead from logging
+   * and to keep volume down.
+   */
   Info = 4,
+  /**
+   * Be silent unless when there is an error or a warning.
+   */
   Warn = 5,
+  /**
+   * Be silent unless when there is an error.
+   */
   Error = 6,
+  /**
+   * Be silent unless when an error is fatal.
+   */
   Fatal = 7,
+  /**
+   * Be silent.
+   */
   Off = 8,
 }
 
+/**
+ * The category to receive log messages for. The {@link LogLevel} will
+ * always be set for a specific category. Setting the log level on one
+ * category, will automatically set the log level for any subcategory.
+ * @note
+ * When debugging, you might not need log messages from everything. To narrow
+ * this scope, log events can be grouped by category.
+ */
+export enum LogCategory {
+  /**
+   * Include logs from all categories. Subcategories are {@link LogCategory.Storage},
+   * {@link LogCategory.Sync}, {@link LogCategory.App}, and {@link LogCategory.SDK}.
+   */
+  Realm = "Realm",
+  /**
+   * Log database mutations and query operations.
+   * Subcategories are {@link LogCategory.Transaction}, {@link LogCategory.Object},
+   * {@link LogCategory.Query}, and {@link LogCategory.Notification}.
+   */
+  Storage = "Realm.Storage",
+  /**
+   * Log when creating, advancing, and committing transactions.
+   */
+  Transaction = "Realm.Storage.Transaction",
+  /**
+   * Log query operations.
+   */
+  Query = "Realm.Storage.Query",
+  /**
+   * Log database mutations.
+   */
+  Object = "Realm.Storage.Object",
+  /**
+   * Log notifications of changes to the database.
+   */
+  Notification = "Realm.Storage.Notification",
+  /**
+   * Log activity related to Atlas Device Sync.
+   * Subcategories are {@link LogCategory.Client} and {@link LogCategory.Server}.
+   */
+  Sync = "Realm.Sync",
+  /**
+   * Log activity related to Atlas Device Sync client operations.
+   * Subcategories are {@link LogCategory.Session}, {@link LogCategory.Changeset},
+   * {@link LogCategory.Network}, and {@link LogCategory.Reset}.
+   */
+  Client = "Realm.Sync.Client",
+  /**
+   * Log connection level activity.
+   */
+  Session = "Realm.Sync.Client.Session",
+  /**
+   * Log when receiving, uploading, and integrating changesets.
+   */
+  Changeset = "Realm.Sync.Client.Changeset",
+  /**
+   * Log low level network activity.
+   */
+  Network = "Realm.Sync.Client.Network",
+  /**
+   * Log client reset operations.
+   */
+  Reset = "Realm.Sync.Client.Reset",
+  /**
+   * Log activity related to Atlas Device Sync server operations.
+   */
+  Server = "Realm.Sync.Server",
+  /**
+   * Log activity at the Atlas App level.
+   */
+  App = "Realm.App",
+  /**
+   * Log activity at the SDK level.
+   */
+  SDK = "Realm.SDK",
+}
```

Comment:
I feel the flattening of the namespace here is a bit misleading and would potentially result in future problems if two categories start using the same sub-category names. I.e. what would happen if core introduce a  `Realm.App.Network` category?

I'd rather that we keep this close to the Core API, use a string-with-dot notation and type this as a union of strings instead:

```suggestion
export type LogCategory = "Realm" | "Realm.Storage" | "Realm.Storage.Transaction" /* ... */;
```

The only drawback is that it doesn't seem to be possible [to document this using a tsdoc string](https://github.com/microsoft/tsdoc/issues/164). But we could probably document the different values on the `LogCategory` type itself or the individual methods taking the value 🤔